datasetinsights/io/tracker/mzflow.py

```diff
@@ -0,0 +1,84 @@
+import os
+
+import mlflow
+from google.auth.transport.requests import Request
+from google.oauth2 import id_token
+
+MLFLOW = "mlflow"
+CLIENT_ID = "tracking_client_id"
+HOST_ID = "tracking_host_id"
+NAME = "experiment_name"
+DEFAULT_NAME = "datasetinsights"
+
+
+class MLFlowTracker:
+    """MlFlow tracker class, responsible for creating singleton object,
+    refresh token and get instance .
+
+    Attributes:
+        __instance: holds singleton instance
+        __client_id: holds ml tracking server client id
+
+    """
+
+    __instance = None
+    __client_id = None
+
+    def __init__(self, client_id):
+        """Virtually private constructor.
+
+        Args:
+            client_id : MLFlow tracking server client id
+        """
+        if MLFlowTracker.__instance:
+            raise Exception("This class is a singleton!")
+        else:
+            if client_id:
+                self.__client_id = client_id
+            MLFlowTracker.__instance = self
+
+    @staticmethod
+    def get_instance(client_id=None):
+
+        """Static instance access method.
+
+        Args:
+            client_id : MLFlow tracking server client id
+        """
+
+        if not MLFlowTracker.__instance:
+            MLFlowTracker(client_id)
+        return MLFlowTracker.__instance
+
+    @staticmethod
+    def get_tracker(config):
+
+        """get instance of tracker class.
+
+        Args:
+            config : config object, holds server details
+        """
+        client_id = config.get(MLFLOW).get(CLIENT_ID, None)
+        host_id = config.get(MLFLOW).get(HOST_ID, None)
+        exp_name = config.get(MLFLOW).get(NAME, DEFAULT_NAME)
+        if host_id:
+            if client_id:
+                MLFlowTracker.refresh_token(client_id)
+            mlflow.set_tracking_uri(host_id)
+            mlflow.set_experiment(experiment_name=exp_name)
+        config.pop(MLFLOW, None)
+        return mlflow
+
+    @staticmethod
+    def refresh_token(client_id=None):
```

Comment:
Does this method tries to fetch token based on GCP credential? What's the desired behavior if client_id is not provided?

Reply:
Yes, service account need to set in environment variable. If clientId is not there then you will get following error
- json.decoder.JSONDecodeError: Expecting value: line 2 column 1 (char 1) when you initialize mlflow :mlflow.set_experiment(experiment_name=exp_name)

Reply:
what's the behavior in the estimator though? it should catch this exception and log a warning since the job can still run just without tracking

Reply:
- if client_id is not provided and host is there:
1. If this run is executed by unity AI team then client id is required, in this it will throw exception and NullTracker will be initiated.
2. If this run is executed by non unity employ, it will try to connect to given tracking host if connected then will return mlflow, if any exception then NullTracker.

- In both the cases estimator will execute.